server/src/main/java/org/elasticsearch/search/fetch/QueryFetchSearchResult.java

```diff
@@ -26,17 +26,21 @@ public final class QueryFetchSearchResult extends SearchPhaseResult {
     private final FetchSearchResult fetchResult;
     private final RefCounted refCounted;
 
+    public static QueryFetchSearchResult of(QuerySearchResult queryResult, FetchSearchResult fetchResult) {
+        // We're acquiring a copy, we should incRef it
+        queryResult.incRef();
+        fetchResult.incRef();
+        return new QueryFetchSearchResult(queryResult, fetchResult);
+    }
+
     public QueryFetchSearchResult(StreamInput in) throws IOException {
         // These get a ref count of 1 when we create them, so we don't need to incRef here
         this(new QuerySearchResult(in), new FetchSearchResult(in));
     }
 
-    public QueryFetchSearchResult(QuerySearchResult queryResult, FetchSearchResult fetchResult) {
+    private QueryFetchSearchResult(QuerySearchResult queryResult, FetchSearchResult fetchResult) {
```

Comment:
This constructor was obviously buggy for the case of reading from the wire where we must not increment the ref count again after instantiating the results objects. Because of the obvious signature conflict and because I didn't want to duplicate the ref counted lambda, went for a simple static constructor here.